### PR TITLE
Locale independent \parse_url() wrapper for UTF-8 strings

### PR DESCRIPTION
**Pull request by alexeyshockov**
Reopening #269 to the right branch

### CODE REVIEW

**Comment by GrahamCampbell** [comment]
Ping @Nyholm for a review. :)

File: src/Uri.php [hunk 1]
```diff
@@ -611,7 +611,8 @@ private function filterHost($host)
             throw new \InvalidArgumentException('Host must be a string');
         }
 
-        return strtolower($host);
+        // strtolower() simply breaks internationalized domain names (depends on the locale selected)
+        return function_exists('mb_strtolower') ? mb_strtolower($host, 'UTF-8') : $host;
```
Comment (GrahamCampbell) [inline src/Uri.php:615]: Do we want to require the symfony polyfil or suggest ext-mb?

Comment (GrahamCampbell) [reply src/Uri.php:615]: Should we hard fail if the string contains non-asci characters and the mb extension or poylfil is not available?

Comment (Nyholm) [reply src/Uri.php:615]: We can just replace this with: 

```
\strtr($host, 'ABCDEFGHIJKLMNOPQRSTUVWXYZ', 'abcdefghijklmnopqrstuvwxyz')
```

File: src/Uri.php [hunk 2]
```diff
@@ -757,4 +758,39 @@ private function validateState()
...
+     *
+     * On the other hand, cURL understands IDN correctly only when UTF-8 locale is
+     * configured ("C.UTF-8", "en_US.UTF-8", etc.).
+     *
+     * @see https://bugs.php.net/bug.php?id=52923
+     * @see https://www.php.net/manual/en/function.parse-url.php#114817
+     * @see https://curl.haxx.se/libcurl/c/CURLOPT_URL.html#ENCODING
+     *
+     * @return array|false
+     * @internal
```
Comment (GrahamCampbell) [inline src/Uri.php:776]: No need to mark private functions as internal. They already are. :)

File: src/functions.php [hunk 3]
```diff
@@ -841,7 +841,7 @@ function _parse_request_uri($path, array $headers)
 }
 
 /**
- * Get a short summary of the message body
+ * Get a short summary of the message body.
```
Comment (GrahamCampbell) [inline src/functions.php:844]: Unrelated change.

File: src/Uri.php [hunk 4]
```diff
@@ -640,7 +641,7 @@ private function filterPort($port)
     /**
      * @param UriInterface $uri
      * @param array        $keys
-     *
```
Comment (GrahamCampbell) [inline src/Uri.php:643]: Unrelated change.

**Review by GrahamCampbell — CHANGES_REQUESTED** [review]
